tool/tsh/putty_config_windows.go

```diff
@@ -0,0 +1,384 @@
+
+// ints
+const puttyDefaultSSHPort = 3022
+const puttyDefaultProxyPort = 0 // no need to set the proxy port as it's abstracted by `tsh proxy ssh`
+
+// dwords
+const puttyDwordPresent = `00000001`
+const puttyDwordProxyMethod = `00000005`    // run a local command
+const puttyDwordProxyLogToTerm = `00000002` // only until session starts
+const puttyPermitRSASHA1 = `00000000`
+const puttyPermitRSASHA256 = `00000001`
+const puttyPermitRSASHA512 = `00000001`
+
+// despite the strings/ints in struct, these are stored in the registry as DWORDs
+type puttyRegistrySessionDwords struct {
+	Present        string // dword
+	PortNumber     int    // dword
+	ProxyPort      int    // dword
+	ProxyMethod    string // dword
+	ProxyLogToTerm string // dword
+}
+
+type puttyRegistrySessionStrings struct {
+	Hostname            string
+	Protocol            string
+	ProxyHost           string
+	ProxyUsername       string
+	ProxyPassword       string
+	ProxyTelnetCommand  string
+	PublicKeyFile       string
+	DetachedCertificate string
+	UserName            string
+}
+
+// addPuTTYSession adds a PuTTY session for the given host/port to the Windows registry
+func addPuTTYSession(proxyHostname string, hostname string, port int, login string, ppkFilePath string, certificateFilePath string, commandToRun string, leafClusterName string) (bool, error) {
+	puttySessionName := fmt.Sprintf(`%v%%20(proxy:%v)`, hostname, proxyHostname)
+	if leafClusterName != "" {
+		puttySessionName = fmt.Sprintf(`%v%%20(leaf:%v,proxy:%v)`, hostname, leafClusterName, proxyHostname)
+	}
+	registryKey := fmt.Sprintf(`%v\%v`, puttyRegistrySessionsKey, puttySessionName)
+	sessionDwords := puttyRegistrySessionDwords{}
+	sessionStrings := puttyRegistrySessionStrings{}
+
+	// if the port passed is 0, this means "use server default" so we override it to 3022
+	if port == 0 {
+		port = puttyDefaultSSHPort
+	}
+
+	sessionDwords = puttyRegistrySessionDwords{
+		Present:        puttyDwordPresent,
+		PortNumber:     port,
+		ProxyPort:      puttyDefaultProxyPort,
+		ProxyMethod:    puttyDwordProxyMethod,
+		ProxyLogToTerm: puttyDwordProxyLogToTerm,
+	}
+
+	sessionStrings = puttyRegistrySessionStrings{
+		Hostname:            hostname,
+		Protocol:            puttyProtocol,
+		ProxyHost:           proxyHostname,
+		ProxyUsername:       login,
+		ProxyPassword:       "",
+		ProxyTelnetCommand:  commandToRun,
+		PublicKeyFile:       ppkFilePath,
+		DetachedCertificate: certificateFilePath,
+		UserName:            login,
+	}
+
+	// now check for and create the individual session key
+	pk, err := getRegistryKey(registryKey)
+	if err != nil {
+		return false, trace.Wrap(err)
+	}
+	defer pk.Close()
+
+	// write dwords
+	if ok, err := registryWriteDword(pk, "Present", sessionDwords.Present); !ok {
+		return false, trace.Wrap(err)
+	}
+	if ok, err := registryWriteDword(pk, "PortNumber", fmt.Sprintf("%v", sessionDwords.PortNumber)); !ok {
```

Comment:
Is it okay to leave the registry in a half-updated state if one of these writes fails, or should you try to clean up after yourself?

Reply:
As there is a possibility that we might be updating a pre-existing PuTTY session (overwriting the named values and leaving everything else the same), I'm loath to try and delete the key. Perhaps I could return a flag from that function that indicates whether we created the key, and if we did, try to delete it here before returning the error. There's a bit of a possibility of a cascading failure, though.